src/Orleans/Serialization/TypeUtilities.cs

```diff
@@ -80,6 +80,11 @@ internal static bool IsOrleansShallowCopyable(this Type t)
             return false;
         }
 
+        private static bool IsTypeFieldsShallowCopyable(TypeInfo typeInfo)
+        {
+            return typeInfo.GetFields().All(f => !(f.FieldType.IsEquivalentTo(typeInfo)) && IsOrleansShallowCopyable(f.FieldType));
```

Comment:
According to tests in the corefx repository: https://github.com/dotnet/corefx/blob/master/src/System.Reflection/tests/TypeInfo/TypeInfo_IsEquivalentTo.cs#L17 
this is compatible with .NET core.